h2o-genmodel/src/main/java/hex/genmodel/AbstractMojoWriter.java

```diff
@@ -0,0 +1,247 @@
+    tmpfile = null;
+  }
+
+
+  //--------------------------------------------------------------------------------------------------------------------
+  // Private
+  //--------------------------------------------------------------------------------------------------------------------
+
+  protected void writeTo(ZipOutputStream zos) throws IOException {
+    writeTo(zos, "");
+  }
+
+  public final void writeTo(ZipOutputStream zos, String zipDirectory) throws IOException {
+    initWriting(zos, zipDirectory);
+    addCommonModelInfo();
+    writeModelData();
+    writeModelInfo();
+    writeDomains();
+    writeExtraInfo();
+  }
+
+  protected void writeExtraInfo() throws IOException {
+    // nothing by default
+  }
+
+  private void initWriting(ZipOutputStream zos, String targetdir) {
+    this.zos = zos;
+    this.targetdir = targetdir;
+  }
+
+  private void addCommonModelInfo() throws IOException {
+    int n_categoricals = 0;
+    for (String[] domain : model.scoringDomains())
+      if (domain != null)
+        n_categoricals++;
+
+    writekv("h2o_version", model.projectVersion());
+    writekv("mojo_version", mojoVersion());
+    writekv("license", "Apache License Version 2.0");
+    writekv("algo", model.algoName().toLowerCase());
+    writekv("algorithm", model.algoFullName());
+    writekv("endianness", ByteOrder.nativeOrder());
+    writekv("category", model.getModelCategory());
+    writekv("uuid", model.uuid());
+    writekv("supervised", model.isSupervised());
+    writekv("n_features", model.nfeatures());
+    writekv("n_classes", model.nclasses());
+    writekv("n_columns", model.columnNames().length);
+    writekv("n_domains", n_categoricals);
+    writekv("balance_classes", model.balanceClasses());
+    writekv("default_threshold", model.defaultThreshold());
+    writekv("prior_class_distrib", Arrays.toString(model.priorClassDist()));
+    writekv("model_class_distrib", Arrays.toString(model.modelClassDist()));
+    writekv("timestamp", model.timestamp());
+  }
+
+  /**
+   * Create the model.ini file containing 3 sections: [info], [columns] and [domains]. For example:
+   * [info]
+   * algo = Random Forest
+   * n_trees = 100
+   * n_columns = 25
+   * n_domains = 3
+   * ...
+   * h2o_version = 3.9.10.0
+   * <p>
+   * [columns]
+   * col1
+   * col2
+   * ...
+   * <p>
+   * [domains]
+   * 5: 13 d000.txt
+   * 6: 7 d001.txt
+   * 12: 124 d002.txt
+   * <p>
+   * Here the [info] section lists general model information; [columns] is the list of all column names in the input
+   * dataframe; and [domains] section maps column numbers (for categorical features) to their domain definition files
+   * together with the number of categories to be read from that file.
+   */
+  private void writeModelInfo() throws IOException {
+    startWritingTextFile("model.ini");
+    writeln("[info]");
+    for (Map.Entry<String, String> kv : lkv.entrySet()) {
+      writeln(kv.getKey() + " = " + kv.getValue());
```

Comment:
StringBuilder ?

Reply:
Or a writeln() version with internal stringBuilder and writeln(String...args) array as arguments :)

Reply:
Java compiler will automatically convert the concatenation to StringBuilder in this case.

      47: new           #11                 // class java/lang/StringBuilder
      50: dup
      51: invokespecial #12                 // Method java/lang/StringBuilder."<init>":()V
      54: aload_2
      55: invokeinterface #102,  1          // InterfaceMethod java/util/Map$Entry.getKey:()Ljava/lang/Object;
      60: checkcast     #103                // class java/lang/String
      63: invokevirtual #14                 // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      66: ldc           #104                // String  =
      68: invokevirtual #14                 // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      71: aload_2
      72: invokeinterface #105,  1          // InterfaceMethod java/util/Map$Entry.getValue:()Ljava/lang/Object;
      77: checkcast     #103                // class java/lang/String
      80: invokevirtual #14                 // Method java/lang/StringBuilder.append:(Ljava/lang/String;)Ljava/lang/StringBuilder;
      83: invokevirtual #15                 // Method java/lang/StringBuilder.toString:()Ljava/lang/String;

I will implement your second suggestion, thank you.